### PR TITLE
les: implement server priority API

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR implements the LES server API. These are the functions provided for server capacity/balance/priority management (see descriptions in api.go):

```
func (api *PrivateLightServerAPI) ServerInfo() map[string]interface{} {
func (api *PrivateLightServerAPI) ClientInfo(ids []enode.ID) map[enode.ID]map[string]interface{} {
func (api *PrivateLightServerAPI) PriorityClientInfo(start, stop enode.ID, maxCount int) map[enode.ID]map[string]interface{} {
func (api *PrivateLightServerAPI) UpdateBalance(id enode.ID, value int64, add bool, meta string) error {
func (api *PrivateLightServerAPI) SetClientParams(ids []enode.ID, params map[string]interface{}) error {
func (api *PrivateLightServerAPI) SetDefaultParams(params map[string]interface{}) error {
func (api *PrivateLightServerAPI) SubscribeEvent(ctx context.Context) (*rpc.Subscription, error) {
```

A note on the "meta" parameter of balance update: this field is stored and updated together with the client balance. It can be used for multiple things:
- remember the last transaction already added to the balance
- store some meta-info about the balance, for example implement token expiration
- tag the client as part of a special subgroup
- store client balance in multiple currencies

The point is that the client pool and the balance tracker only needs to know about a scalar balance/priority value, all other information about clients is handled and interpreted by higher layers so the server code does not need to be more complicated.

### CODE REVIEW

File: les/api.go [hunk 1]
```diff
@@ -17,16 +17,408 @@
...
+// SetClientParams sets client parameters for all clients listed in the ids list or matching the given tags
+func (api *PrivateLightServerAPI) SetClientParams(ids []enode.ID, params map[string]interface{}) error {
+	updateFactorsWithParam := func(name string, value interface{}, posFactors, negFactors *priceFactors) (bool, error) {
+		var err error
+		switch name {
+		case "pricing/timeFactor":
+			if val, ok := value.(float64); ok && val >= 0 {
+				posFactors.timeFactor = val / 1000000000
+			} else {
+				err = errInvalidValue
```
Comment (holiman) [inline les/api.go]: Wouldn't it be nicer to say which value was invalid? `err = errors.New("invalid value for %v", name)`

File: les/api.go [hunk 2]
```diff
@@ -17,16 +17,408 @@
...
+				if u, err = updateFactorsWithParam(name, value, &dummy, &dummy); u {
+					err = errClientNotConnected
+				}
+			}
+			if !u && err == nil {
+				switch name {
+				case "capacity":
+					if client != nil {
+						if capacity, ok := value.(float64); ok && (capacity == 0 || uint64(capacity) >= api.server.minCapacity) {
+							err = api.server.clientPool.setCapacity(client, uint64(capacity))
```
Comment (holiman) [inline les/api.go]: this is an `if` in an `if` in a  `case` in a `switch` in a `for` in an anonymous `func`.  Can this be rewritten for clarity?

Comment (zsfelfoldi) [reply les/api.go]: yes, done

File: les/api.go (hunk 1 above)
Comment (zsfelfoldi) [reply les/api.go]: done

File: les/clientpool_test.go [hunk 3]
```diff
@@ -102,9 +104,11 @@ func testClientPool(t *testing.T, connLimit, clientCount, paidCount int, randomD
 
 		if tickCounter == testClientPoolTicks/4 {
 			// give a positive balance to some of the peers
-			amount := uint64(testClientPoolTicks / 2 * 1000000000) // enough for half of the simulation period
+			amount := int64(testClientPoolTicks / 2 * 1000000000) // enough for half of the simulation period
```
Comment (ligi) [inline les/clientpool_test.go]: why the change from uint to int here?

Comment (ligi) [inline les/clientpool_test.go]: I think numbers with so many zeroes are bad in code - wonder if go can do something like 1_000_000_000 or alternative 10**9

File: les/clientpool.go [hunk 4]
```diff
@@ -509,65 +622,108 @@ func (f *clientPool) getPosBalance(id enode.ID) *posBalance {
...
 
-// addBalance updates the positive balance of a client.
-// If setTotal is false then the given amount is added to the balance.
-// If setTotal is true then amount represents the total amount ever added to the
-// given ID and positive balance is increased by (amount-lastTotal) while lastTotal
-// is updated to amount. This method also allows removing positive balance.
-func (f *clientPool) addBalance(id enode.ID, amount uint64, setTotal bool) {
-	f.lock.Lock()
-	defer f.lock.Unlock()
+// getPosBalanceIDs returns a lexicographically ordered list of IDs of accounts
```
Comment (ligi) [inline les/clientpool.go]: "of IDs of accounts" -> "of IDs from accounts"

Comment (zsfelfoldi) [reply les/clientpool.go]: fixed

File: les/clientpool_test.go (hunk 3 above)
Comment (zsfelfoldi) [reply les/clientpool_test.go]: its real meaning is nanoseconds per second so I figured I can use time.Second here

File: les/clientpool.go [hunk 5]
```diff
@@ -509,65 +622,108 @@ func (f *clientPool) getPosBalance(id enode.ID) *posBalance {
...
+	id                enode.ID
+	value, lastStored uint64
+	meta              string
+	queueIndex        int // position in posBalanceQueue
 }
 
 // EncodeRLP implements rlp.Encoder
 func (e *posBalance) EncodeRLP(w io.Writer) error {
-	return rlp.Encode(w, []interface{}{e.value, e.lastTotal})
+	return rlp.Encode(w, []interface{}{e.value, e.meta})
```
Comment (rjl493456442) [inline les/clientpool.go]: The database schema change will definitely introduce some troubles. e.g. we should consider backward-compatibility

File: les/server.go [hunk 6]
```diff
@@ -49,9 +49,9 @@ type LesServer struct {
 	servingQueue *servingQueue
 	clientPool   *clientPool
 
-	freeCapacity uint64 // The minimal client capacity used for free client.
-	threadsIdle  int    // Request serving threads count when system is idle.
-	threadsBusy  int    // Request serving threads count when system is busy(block insertion).
+	minCapacity, maxCapacity, freeCapacity uint64
```
Comment (rjl493456442) [inline les/server.go]: Is it worthwhile to maintain `minCapacity ` and `maxCapacity `? Basically these two fields are the limit of capacity and we can re-calculate them via freeCapacity

File: les/api.go [hunk 7]
```diff
@@ -17,16 +17,400 @@
...
+
+// NewPrivateDebugAPI creates a new LES light server debug API.
+func NewPrivateDebugAPI(server *LesServer) *PrivateDebugAPI {
+	return &PrivateDebugAPI{
+		server: server,
+	}
+}
+
+// FreezeClient forces a temporary client freeze which normally happens when the server is overloaded
+func (api *PrivateDebugAPI) FreezeClient(id enode.ID) error {
```
Comment (rjl493456442) [inline les/api.go]: It's a part of manual management APIs. The main goal of these APIs is the operator can control the clients more easily.

If so, why not offer more? like 

`FreezeClient(id enode.ID, duration time.Duration)`: operator can choose to stop to provide service to this client for that long time

`UnfreezeClient(id enode.ID)`

`DropClient(id enode.ID)`

but we can definitely add more APIs later.

File: les/api.go [hunk 8]
```diff
@@ -17,16 +17,400 @@
...
+					err = errValue()
+				}
+			} else {
+				err = errClientNotConnected
+			}
+		}
+
+		processed := true
+		switch name {
+		case "pricing/timeFactor":
```
Comment (rjl493456442) [inline les/api.go]: TBH, is it might too technical deep for operator to set these configurations?

I guess all we need to do is offer the unit of resource client uses, then all the operator needed is just a price per unit.

Comment (zsfelfoldi) [reply les/api.go]: This is something I would like to leave open for experimentation. Usually a script is going to set these parameters and it is pretty easy to hardcode the ratio there.

File: les/api.go (hunk 7 above)
Comment (zsfelfoldi) [reply les/api.go]: Freezing is something that should usually happen automatically, I only added this call to the debug API for testing purpose and right now we don't need any other debug calls. We can add them later if there is a reason. Also, freezing has a short default interval and it is not a good practice to freeze for longer periods so I don't want to encourage people doing that.

File: les/api.go [hunk 9]
```diff
@@ -17,16 +17,400 @@
...
+func (api *PrivateLightServerAPI) clientInfo(c *clientInfo, id enode.ID) map[string]interface{} {
+	info := make(map[string]interface{})
+	if c != nil {
+		info["isConnected"] = true
+		info["capacity"] = c.capacity
+		info["pricing/balance"], info["pricing/negBalance"] = c.balanceTracker.getBalance(mclock.Now())
+		info["pricing/balanceMeta"] = c.balanceMetaInfo
+	} else {
+		info["isConnected"] = false
+		pb := api.server.clientPool.getPosBalance(id)
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/api.go]: What do you mean by "ignore"? If it is explicitly requested by ID then I think it is ok to return a result showing no connection and no balance. `ClientInfo` with empty `ids` field won't return these because they are not connected, neither does `PriorityClientInfo` because they don't have a balance so in this sense they are already treated as non-existent.

Priority status is obvious from the `balance` field but I guess we can return it explicitly. Returning connection time is a good idea, I'll add that too.

File: les/api.go [hunk 10]
```diff
@@ -17,16 +17,400 @@
...
+	res["minimumCapacity"] = api.server.minCapacity
+	res["maximumCapacity"] = api.server.maxCapacity
+	res["freeClientCapacity"] = api.server.freeCapacity
+	res["totalCapacity"], res["totalConnectedCapacity"], res["priorityConnectedCapacity"] = api.server.clientPool.capacityInfo()
+	return res
+}
+
+// ClientInfo returns information about clients listed in the ids list or matching the given tags
+func (api *PrivateLightServerAPI) ClientInfo(ids []enode.ID) map[enode.ID]map[string]interface{} {
+	res := make(map[enode.ID]map[string]interface{})
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/api.go]: `ClientInfo` always returns a limited number of clients (either the ones explicitly requested or all connected ones, which is also limited). So I don't think we need start-stop there.
`IterateFreeClient` is an interesting idea, if I understand correctly then it would iterate through the negative balance db and return unconnected free clients who were using the service recently. Start-stop mode would make sense there. Maybe I'll add it after the already existing functions are finished.

File: les/server.go (hunk 6 above)
Comment (zsfelfoldi) [reply les/server.go]: `maxCapacity` does not follow from `freeCapacity` directly. `minCapacity` always equals `freeCapacity` in the current implementation but they are two different concepts and maybe at one point we would like to allow setting `freeCapacity` higher than the absolute minimum.

File: les/clientpool.go [hunk 11]
```diff
@@ -509,65 +622,108 @@ func (f *clientPool) getPosBalance(id enode.ID) *posBalance {
...
+		if len(result) == maxCount {
+			return
+		}
+	}
+	return
+}
 
+// updateBalance updates the balance of a client (either overwrites it or adds to it).
+// It also updates the balance meta info string.
+func (f *clientPool) updateBalance(id enode.ID, amount int64, add bool, meta string) error {
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: Right, I also started to not like this so much, it is removed now.

File: les/api.go [hunk 12]
```diff
@@ -17,16 +17,400 @@
...
+	return api
+}
+
+// ServerInfo returns global server parameters
+func (api *PrivateLightServerAPI) ServerInfo() map[string]interface{} {
+	res := make(map[string]interface{})
+	res["minimumCapacity"] = api.server.minCapacity
+	res["maximumCapacity"] = api.server.maxCapacity
+	res["freeClientCapacity"] = api.server.freeCapacity
+	res["totalCapacity"], res["totalConnectedCapacity"], res["priorityConnectedCapacity"] = api.server.clientPool.capacityInfo()
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/api.go]: I am not sure... it does not matter too much because any two of the three values are sufficient to calculate the third one but I think the total connected value is the most interesting one of the three so it would feel weird to not return that directly.

File: les/clientpool.go (hunk 5 above)
Comment (zsfelfoldi) [reply les/clientpool.go]: What do you mean? This code is not used yet by anyone so there is no compatibility issue. The `meta` field should be flexible enough so we don't need to add more application specific fields. The script using this API can use the `meta` field for whatever it needs, for example it can store the id of the last processed incoming tx to make sure that everything is added exactly once.

File: les/api.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/api.go]: added `priority` and `connectionTime` fields

File: les/clientpool.go [hunk 13]
```diff
@@ -627,6 +761,37 @@ func (db *nodeDB) delPB(id enode.ID) {
...
 
+// getPosBalanceIDs returns a lexicographically ordered list of IDs of accounts
+// with a positive balance
+func (db *nodeDB) getPosBalanceIDs(start, stop enode.ID, maxCount int) (result []enode.ID) {
+	if maxCount <= 0 {
+		return
+	}
+	it := db.db.NewIteratorWithStart(db.key(start.Bytes(), false))
+	defer it.Release()
+	for i := len(stop[:]) - 1; i >= 0; i-- {
```
Comment (rjl493456442) [inline les/clientpool.go]: It seems unnecessarily complicated.

Why not extract the id from `iter.Key` every time and compare with `stop` enode.ID?

Anyway, we need to extract the ID to return.

File: les/clientpool.go [hunk 14]
```diff
@@ -374,14 +443,22 @@ func (f *clientPool) balanceExhausted(id enode.ID) {
...
 	if c.capacity != f.freeClientCap {
 		f.connectedCap += f.freeClientCap - c.capacity
 		totalConnectedGauge.Update(int64(f.connectedCap))
 		c.capacity = f.freeClientCap
 		c.peer.updateCapacity(c.capacity)
 	}
-	f.ndb.delPB(id)
+	pb := f.ndb.getOrNewPB(id)
+	pb.value = 0
+	f.ndb.setPB(id, pb)
```
[the remark this answers is not in the document]
Comment (rjl493456442) [reply les/clientpool.go]: Ah, I see the modification in `nodedb. setPB` why not just delete it? If the balance is 0, then the meta doesn't have meaning anymore I think.

File: les/clientpool.go [hunk 15]
```diff
@@ -460,44 +581,53 @@ func (f *clientPool) addBalance(id enode.ID, amount uint64, setTotal bool) {
 			c.balanceTracker.setBalance(pb.value, negBalance)
 			if !c.priority && pb.value > 0 {
 				c.priority = true
+				if c.priority {
+					f.priorityConnected += c.capacity
+				}
 				c.balanceTracker.addCallback(balanceCallbackZero, 0, func() { f.balanceExhausted(id) })
 			}
+			c.balanceMetaInfo = meta
```
Comment (rjl493456442) [inline les/clientpool.go]: In this defer function, even the error is returned, meta of `clientInfo` will be updated.

File: les/clientpool.go [hunk 16]
```diff
@@ -611,6 +741,10 @@ func (db *nodeDB) getOrNewPB(id enode.ID) posBalance {
 }
 
 func (db *nodeDB) setPB(id enode.ID, b posBalance) {
+	if b.value == 0 && len(b.meta) == 0 {
```
Comment (rjl493456442) [inline les/clientpool.go]: Please use `nodeDB.delPB` directly.

File: les/api.go [hunk 17]
```diff
@@ -17,16 +17,407 @@
...
+				continue loop
+			}
+		}
+		if processed {
+			continue loop
+		}
+		switch name {
+		case "capacity":
+			if client != nil {
+				if capacity, ok := value.(float64); ok && (capacity == 0 || uint64(capacity) >= api.server.minCapacity) {
```
Comment (rjl493456442) [inline les/api.go]: Why `capacity == 0` is allowed?

File: les/api.go [hunk 18]
```diff
@@ -17,16 +17,407 @@
...
+	if client != nil {
+		posFactors, negFactors = &client.posFactors, &client.negFactors
+	}
+	defParams := id == enode.ID{}
+loop:
+	for name, value := range params {
+		errValue := func() error {
+			return fmt.Errorf("invalid value for parameter '%s'", name)
+		}
+		setFactor := func(v *float64) {
```
Comment (rjl493456442) [inline les/api.go]: This function seems weird. If `posFactors == nil` why we will return the error `errClientNotConnected `?

File: les/api.go [hunk 19]
```diff
@@ -17,16 +17,407 @@
...
+		case "pricing/negative/timeFactor":
+			setFactor(&negFactors.timeFactor)
+		case "pricing/negative/capacityFactor":
+			setFactor(&negFactors.capacityFactor)
+		case "pricing/negative/requestCostFactor":
+			setFactor(&negFactors.requestFactor)
+		default:
+			processed = false
+			if defParams {
+				err = fmt.Errorf("invalid default parameter '%s'", name)
```
Comment (rjl493456442) [inline les/api.go]: If there is any error, why not return directly?

File: les/api.go (hunk 17 above)
Comment (zsfelfoldi) [reply les/api.go]: Right, it shouldn't be. It was left there from an early version where capacity=0 was actually a free client with freeClientCapacity assigned. Now it is always the actual capacity so I fixed it now.

File: les/api.go (hunk 18 above)
Comment (zsfelfoldi) [reply les/api.go]: This whole function was a mess :( Earlier setParams handled balance updates so it made sense to call it for disconnected clients. Now it does not so I simplified the logic (which was buggy anyways). I will check the whole code again for stuff like this after finishing all changes.

File: les/api.go (hunk 19 above)
Comment (zsfelfoldi) [reply les/api.go]: This logic actually might make sense because if we are setting many parameters and returning in the middle of it because of an error then it is not clear which parameters have been changed. This way it is clear that every valid parameter change has happened.
Checking for all possible errors before changing anything would be another logical alternative but that would probably make the code unnecessarily complex.

File: les/clientpool.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: It might have a meaning, it was a deliberate change to keep it if not empty. One example: if meta stores an ID of the last processed incoming payment. If the balance of the client gets exhausted but a new payment arrives around the same time (and also we have been recently restarted so we are not sure which payments are processed in the current state of the node db) then it is useful to know that the exhausted balance did not include the last payment so we can add it.

File: les/clientpool.go [hunk 20]
```diff
@@ -460,44 +581,53 @@ func (f *clientPool) addBalance(id enode.ID, amount uint64, setTotal bool) {
...
+		if amount > maxBalance || pb.value > maxBalance-uint64(amount) {
+			return errBalanceOverflow
 		}
-		pb.lastTotal = amount
+		pb.value += uint64(amount)
 	} else {
-		pb.value += amount
-		pb.lastTotal += amount
+		if uint64(-amount) > pb.value {
+			pb.value = 0
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: Right, at least some feedback would be necessary.

File: les/clientpool.go [hunk 21]
```diff
@@ -460,44 +581,53 @@ func (f *clientPool) addBalance(id enode.ID, amount uint64, setTotal bool) {
 			c.balanceTracker.setBalance(pb.value, negBalance)
 			if !c.priority && pb.value > 0 {
 				c.priority = true
+				if c.priority {
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: right, done

File: les/clientpool.go [hunk 22]
```diff
@@ -240,11 +258,14 @@ func (f *clientPool) connect(peer clientPeer, capacity uint64) bool {
 		capacity = f.freeClientCap
 	}
 	e.capacity = capacity
+	if e.priority {
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: right, fixed. the balance tracker was also started too early, fixed that too.

File: les/clientpool.go (hunk 13 above)
Comment (zsfelfoldi) [reply les/clientpool.go]: Because NewIteratorWithStart is not NewIteratorWithPrefix so this iterator can run out of the balance entry range. Comparing the entire key ensures that we stop as soon as this happens.
Btw is there any way to iterate only in a prefix range but also start from a given key? That would allow us to simplify this routine but I don't know about any easy way to do that.

File: les/clientpool.go [hunk 23]
```diff
@@ -398,6 +475,57 @@ func (f *clientPool) setLimits(totalConn int, totalCap uint64) {
...
+		return nil
+	}
+	if !c.priority {
+		return errNoPriority
+	}
+	oldCapacity := c.capacity
+	c.capacity = capacity
+	f.connectedCap += capacity - oldCapacity
+	f.connectedQueue.Remove(c.queueIndex)
+	f.connectedQueue.Push(c)
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: right, done

File: les/clientpool.go [hunk 24]
```diff
@@ -398,6 +475,57 @@ func (f *clientPool) setLimits(totalConn int, totalCap uint64) {
...
+			}
+			return kick && (f.connectedCap > f.capLimit)
+		})
+		if kick {
+			now := mclock.Now()
+			for _, c := range kickList {
+				f.dropClient(c, now, true)
+			}
+		} else {
+			c.capacity = oldCapacity
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: `LazyQueue` can tell us whether the update is possible and the priority callback gets the priority from `balanceTracker` so it is a lot easier to change and then revert if necessary.

File: les/clientpool.go (hunk 16 above)
Comment (zsfelfoldi) [reply les/clientpool.go]: As I mentioned above, I want to keep entries with non-empty meta tags.

File: les/clientpool.go (hunk 15 above)
Comment (zsfelfoldi) [reply les/clientpool.go]: fixed

File: les/clientpool.go [hunk 25]
```diff
@@ -398,6 +475,57 @@ func (f *clientPool) setLimits(totalConn int, totalCap uint64) {
...
+				f.connectedCap += c.capacity
+				f.connectedQueue.Push(c)
+			}
+			return errNoPriority
+		}
+	}
+	totalConnectedGauge.Update(int64(f.connectedCap))
+	f.priorityConnected += capacity - oldCapacity
+	c.peer.updateCapacity(c.capacity)
+	return nil
```
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: The problem with your version is that if `c.balanceTracker` is not updated with the new capacity and its priority is not updated in `f.connectedQueue` then `c` is still in its old place in the queue and `MultiPop` will return a wrong order. The question here is whether `c` can stay in the queue with its _new_ priority (which is affected by the capacity). So I think there is no easy way around updating and then reverting it if necessary. Also, I don't think it is more complicated, we just have to take care of properly reverting the original state if necessary before releasing the lock.

[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: I changed the error message to "priority too low to raise capacity" because that describes both cases perfectly. Zero priority is obviously not enough. If there is "no space" to raise capacity then that is also an issue of low priority because with a higher priority other clients could be kicked out.
I am thinking about returning the missing amount of balance in order to successfully change the capacity so that the client can decide it easier if it wants to buy the necessary amount of tokens.

File: les/clientpool.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (rjl493456442) [reply les/clientpool.go]: Yes, it makes sense. 

So why not add another field for payment relative? I think we should store the last applied payment sequence number for all paid clients. `BalanceMeta`is only for the extra information?

File: les/api.go (hunk 19 above)
Comment (rjl493456442) [reply les/api.go]: But the error will be overwritten over and over again.  If if we have 2 invaid parameters, it's also unclear

File: les/clientpool.go [hunk 26]
```diff
@@ -301,6 +318,9 @@ func (f *clientPool) connect(peer clientPeer, capacity uint64) bool {
 	totalConnectedGauge.Update(int64(f.connectedCap))
 	clientConnectedMeter.Mark(1)
 	log.Debug("Client accepted", "address", freeID)
+	if f.eventHook != nil {
```
Comment (rjl493456442) [inline les/clientpool.go]: It's dangerous to call `eventHook` directly here. Because it involves network IO which can block the entire function.

Let's add a TODO: change the event notification system to non-blocking mode.

File: les/api.go [hunk 27]
```diff
@@ -17,16 +17,398 @@
...
+
+	event := make(map[string]interface{})
+	event["totalCapacity"], event["totalConnectedCapacity"], event["priorityConnectedCapacity"] = api.server.clientPool.capacityInfo()
+	if client != nil {
+		event["clientEvent"] = clientEvent
+		event["clientId"] = client.id
+		event["clientInfo"] = api.clientInfo(client, client.id)
+	}
+
+	for sub := range api.subs {
```
Comment (rjl493456442) [inline les/api.go]: It involves a concurrent map read and write.

File: les/clientpool.go [hunk 28]
```diff
@@ -398,6 +472,59 @@ func (f *clientPool) setLimits(totalConn int, totalCap uint64) {
...
+			return kick && (f.connectedCap > f.capLimit)
+		})
+		if kick {
+			now := mclock.Now()
+			for _, c := range kickList {
+				f.dropClient(c, now, true)
+			}
+		} else {
+			c.capacity = oldCapacity
+			c.balanceTracker.setCapacity(oldCapacity)
```
Comment (rjl493456442) [inline les/clientpool.go]: So we still need to set back `totalConnected`

**Comment by rjl493456442** [comment]
Btw, I test the new node, dead lock happens.

I  subscribe the server events. The root cause is `sendEvent` will call `api.server.clientPool.capacityInfo()`, which needs lock.

File: les/clientpool.go (hunk 28 above)
Comment (zsfelfoldi) [reply les/clientpool.go]: event subs are removed for now

Comment (zsfelfoldi) [reply les/clientpool.go]: All clients in `kickList` are removed from `connectedCap`, including `c` if `kick` is false. They are added back after resetting `c.capacity = oldCapacity` so `connectedCap` should be correct in the end.

File: les/api.go (hunk 27 above)
Comment (zsfelfoldi) [reply les/api.go]: event subs are removed for now

File: les/clientpool.go (hunk 26 above)
Comment (zsfelfoldi) [reply les/clientpool.go]: event subs are removed for now

File: les/clientpool.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (zsfelfoldi) [reply les/clientpool.go]: I think it might depend on the underlying payment system. Maybe it uses a single sequence number, maybe a block number+hash of the chain where the payments are received... the `meta` field can store any kind of reference and also additional extra info.

File: les/api.go (hunk 19 above)
Comment (zsfelfoldi) [reply les/api.go]: Ok, I changed the error handling. Now it returns on the first error so the caller has to retry until the call returns with no error. Now that adding balance is a separate function and not handled by `setParams`, setting something twice is not a problem.

**Comment by fjl** [comment]
@zsfelfoldi Please document the API on the gh-pages branch. You can use https://geth.ethereum.org/docs/rpc/ns-miner as a guide, just copy https://github.com/ethereum/go-ethereum/blob/gh-pages/docs/_rpc/ns-miner.md to a new file (ns-les.md) and add info about each API method. It would be nice to have a bit of intro text about this API as well, i.e. explain who the API is for and what you can do with with it.
